### PR TITLE
various sysctl-util.c cleanups

### PR DESCRIPTION
**Pull request by poettering**

### CODE REVIEW

**Review by yuwata — APPROVED** [review]
LGTM.

**Comment by poettering** [comment]
hmm, the CI failures are probably relevant

**Comment by yuwata** [comment]
```
======================================================================
FAIL: test_sysctl (__main__.NetworkdNetworkTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/test/test-network/systemd-networkd-tests.py", line 2750, in test_sysctl
    self.assertEqual(read_ipv6_sysctl_attr('dummy98', 'forwarding'), '1')
AssertionError: '0' != '1'
- 0
+ 1
```

**Comment by yuwata** [comment]
```
Setting '/proc/sys/net/ipvipv4/ip_forward' to '1'
dummy98: Cannot turn on IPv4 packet forwarding, ignoring: No such file or directory
dummy98: Cannot configure IPv6 packet forwarding, ignoring: Invalid argument
```

File: src/basic/sysctl-util.c [hunk 1]
```diff
@@ -77,17 +79,21 @@ int sysctl_writef(const char *property, const char *format, ...) {
...
 
-        log_debug("Setting '%s' to '%s'", p, value);
+        if (ifname) {
+                if (!ifname_valid(ifname))
+                        return -EINVAL;
 
-        return write_string_file(p, value, WRITE_STRING_FILE_VERIFY_ON_FAILURE | WRITE_STRING_FILE_DISABLE_BUFFER);
+                p = strjoina("net/", af_to_ipv4_ipv6(af), "/conf/", ifname, "/", property);
+        } else
+                p = strjoina("net/ipv", af_to_ipv4_ipv6(af), "/", property);
```
Comment (yuwata) [inline src/basic/sysctl-util.c]: `"net/ipv"` -> `"net/"`

File: src/basic/sysctl-util.c [hunk 2]
```diff
@@ -118,24 +118,20 @@ int sysctl_read(const char *property, char **ret) {
...
+        if (ifname) {
+                if (!ifname_valid(ifname))
+                        return -EINVAL;
 
-        truncate_nl(value);
-        if (ret)
-                *ret = TAKE_PTR(value);
+                p = strjoina("net/", af_to_ipv4_ipv6(af), "/conf/", ifname, "/", property);
+        } else
+                p = strjoina("net/ipv", af_to_ipv4_ipv6(af), "/", property);
```
Comment (yuwata) [inline src/basic/sysctl-util.c]: Same here. `"net/ipv"` -> `"net/"`

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/basic/sysctl-util.c [hunk 3]
```diff
@@ -77,17 +79,21 @@ int sysctl_writef(const char *property, const char *format, ...) {
...
 
-        p = strjoina("/proc/sys/net/ipv", af == AF_INET ? "4" : "6",
-                     ifname ? "/conf/" : "", strempty(ifname),
-                     property[0] == '/' ? "" : "/", property);
+        if (!IN_SET(af, AF_INET, AF_INET6))
+                return -EAFNOSUPPORT;
 
-        log_debug("Setting '%s' to '%s'", p, value);
+        if (ifname) {
+                if (!ifname_valid(ifname))
```
Comment (yuwata) [inline src/basic/sysctl-util.c]: `"all"` and `"default"` should be also allowed.

File: src/basic/sysctl-util.c [hunk 4]
```diff
@@ -118,24 +118,20 @@ int sysctl_read(const char *property, char **ret) {
...
-                     ifname ? "/conf/" : "", strempty(ifname),
-                     property[0] == '/' ? "" : "/", property);
+        if (!IN_SET(af, AF_INET, AF_INET6))
+                return -EAFNOSUPPORT;
 
-        r = read_full_virtual_file(p, &value, NULL);
-        if (r < 0)
-                return r;
+        if (ifname) {
+                if (!ifname_valid(ifname))
```
Comment (yuwata) [inline src/basic/sysctl-util.c]: Here too.

**Review by yuwata — CHANGES_REQUESTED** [review]

**Comment by poettering** [comment]
force pushed a new version. fixed the issues raised, and added a test to make sure the CI failure from earlier doesn't regress

**Review by yuwata — APPROVED** [review]
LGTM.

File: src/basic/sysctl-util.c [hunk 5]
```diff
@@ -36,33 +38,27 @@ char *sysctl_normalize(char *s) {
...
+        path_simplify(p);
+        if (!path_is_normalized(p))
+                return -EINVAL;
 
-        if (write(fd, value, strlen(value)) < 0)
-                return -errno;
+        log_debug("Setting '%s' to '%s'", p, value);
 
-        return 0;
+        return write_string_file(p, value, WRITE_STRING_FILE_VERIFY_ON_FAILURE | WRITE_STRING_FILE_DISABLE_BUFFER);
```
Comment (floppym) [inline src/basic/sysctl-util.c:61]: We intentionally made sysctl_write() call write() directly in 521251d2757295b6e9df4b51c7cb33929fbd65c4 to resolve #7744. It seems likely that you have re-introduced that problem.

**Pull request merged**